pkg/velerorunner/task.go

```diff
@@ -0,0 +1,382 @@
+	if err != nil {
+		return err
+	}
+	if foundRestore == nil {
+		t.Restore = newRestore
+		client, err := t.getDestinationClient()
+		if err != nil {
+			return err
+		}
+		err = client.Create(context.TODO(), newRestore)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	t.Restore = foundRestore
+	if !t.equalsRestore(newRestore, foundRestore) {
+		t.updateRestore(foundRestore)
+		client, err := t.getDestinationClient()
+		if err != nil {
+			return err
+		}
+		err = client.Update(context.TODO(), foundRestore)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Get whether the two Restores are equal.
+func (t *Task) equalsRestore(a, b *velero.Restore) bool {
+	match := a.Spec.BackupName == b.Spec.BackupName &&
+		*a.Spec.RestorePVs == *b.Spec.RestorePVs
+	return match
+}
+
+// Get an existing Restore on the destination cluster.
+func (t Task) getRestore() (*velero.Restore, error) {
+	client, err := t.getDestinationClient()
+	if err != nil {
+		return nil, err
+	}
+	list := velero.RestoreList{}
+	labels := t.Owner.GetCorrelationLabels()
+	err = client.List(
+		context.TODO(),
+		k8sclient.MatchingLabels(labels),
+		&list)
+	if err != nil {
+		return nil, err
+	}
+	if len(list.Items) > 0 {
+		return &list.Items[0], nil
+	}
+
+	return nil, nil
+}
+
+// Build a Restore as desired for the destination cluster.
+func (t *Task) buildRestore() *velero.Restore {
+	restore := &velero.Restore{
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:       t.Owner.GetCorrelationLabels(),
+			GenerateName: t.Owner.GetName() + "-",
+			Namespace:    VeleroNamespace,
+		},
+	}
+	t.updateRestore(restore)
+	return restore
+}
+
+// Update a Restore as desired for the destination cluster.
+func (t *Task) updateRestore(restore *velero.Restore) {
+	restorePVs := true
+	restore.Spec = velero.RestoreSpec{
+		BackupName: t.Backup.Name,
+		RestorePVs: &restorePVs,
+	}
+}
+
+// Get a Backup that has been replicated by velero on the destination cluster.
+func (t *Task) getReplicaatedBackup() (*velero.Backup, error) {
+	client, err := t.getDestinationClient()
+	if err != nil {
+		return nil, err
+	}
+	list := velero.BackupList{}
+	labels := t.Owner.GetCorrelationLabels()
+	err = client.List(
+		context.TODO(),
+		k8sclient.MatchingLabels(labels),
+		&list)
+	if err != nil {
+		return nil, err
+	}
+	if len(list.Items) > 0 {
+		return &list.Items[0], nil
+	}
+
+	return nil, nil
+}
+
+// Get a client for the source cluster using the cache.
+func (t *Task) getSourceClient() (k8sclient.Client, error) {
+	return t.getClient(t.PlanResources.SrcMigCluster)
+}
+
+// Get a client for the destination cluster using the cache.
+func (t *Task) getDestinationClient() (k8sclient.Client, error) {
+	return t.getClient(t.PlanResources.DestMigCluster)
+}
+
+// Get a client for the cluster using the cache.
+func (t *Task) getClient(cluster *migapi.MigCluster) (k8sclient.Client, error) {
+	if t.ClientCache == nil {
+		t.ClientCache = map[*migapi.MigCluster]k8sclient.Client{}
```

Comment:
What is this client cache?

Reply:
The client cache keeps the `Task` from constantly creating `Client`s.  Reusing the `Client` for the duration of the `Task.Run()` for efficiency especially for remote clients.

Reply:
:+1: